### ISSUE TITLE
Trouble reproducing results from table 2 in paper

### ISSUE THREAD

**Issue by hyukkyukang** [zycdev/AISO#4]
Hello, @zycdev 

I trained the model using preprocessed data from [Google drive](https://drive.google.com/drive/folders/1n38_RIhz5jyZQUFLkp-QUVERhAI7E6nn?usp=sharing). 
However, I was not able to reproduce the performance metrics mentioned in Table 2 of the paper. 
I utilized the large version of the pretrained language model, but the result I obtained is as follows:
```
{
    "em": 0.6013504388926401,
    "f1": 0.722244874910069,
    "prec": 0.7459586201433037,
    "recall": 0.735388557647664
}
```

The performance metrics I achieved are approximately 8% lower than those reported in the paper. <img width="808" alt="image" src="https://user-images.githubusercontent.com/106138715/236726986-968b67ab-6556-421e-9032-d253b6fc763d.png">

Can you please tell me if the results presented in the paper were obtained using the model trained with the command specified [here](https://github.com/zycdev/AISO/blob/main/train_union.py#L55).

Any guidance or assistance in reproducing the paper's results would be greatly appreciated. 

Thank you!

**Comment by hyukkyukang** [comment]
Hello, I've attached my TensorBoard training log below. I would greatly appreciate it if you could review the log and let me know if there are any notable discrepancies when compared to your previous experiments.

**Learning rate:**
<img width="343" alt="image" src="https://user-images.githubusercontent.com/106138715/236728292-db504c5d-8085-4dd9-a4fe-c388267064f2.png">

**Loss:**
<img width="1386" alt="image" src="https://user-images.githubusercontent.com/106138715/236728234-620b8d14-f587-4149-a669-26e9c1f9be3c.png">

**Metric:**
<img width="1373" alt="image" src="https://user-images.githubusercontent.com/106138715/236728273-7efbafbc-3500-4dc9-9d5a-d341e8d7fae5.png">
<img width="1370" alt="image" src="https://user-images.githubusercontent.com/106138715/236729119-49b5049a-75fc-4e35-af90-36f452bb6f2d.png">

**Comment by zycdev** [comment]
Have you tested it in game.ipynb? The evaluation at training time is performed under a random sampling distribution, which is not the same as the observation distribution encountered during actual testing, so the metrics in tensorboard are only used as a proxy for selecting the model.

**Comment by hyukkyukang** [comment]
Hello, thanks for the reply!

Yes, the accuracy I got is based on the game.ipynb (i.e., the observation distribution for HotpotQA Dev. set).

To be more specific, when evaluated with max step 10, the accuracy is:
```
{
    "em": 0.6013504388926401,
    "f1": 0.722244874910069,
    "prec": 0.7459586201433037,
    "recall": 0.735388557647664
}
```
When the model is evaluated with max step 100, the accuracy is:
```
{
    "em": 0.6294395678595543,
    "f1": 0.7560857453192302,
    "prec": 0.7816758087840826,
    "recall": 0.768849493890504,
    "sp_em": 0.5417960837272113,
    "sp_f1": 0.8000359779765738,
    "sp_prec": 0.8267589466576687,
    "sp_recall": 0.7929479759493299,
    "joint_em": 0.399459824442944,
    "joint_f1": 0.6580608787627396,
    "joint_prec": 0.7009365136805664,
    "joint_recall": 0.663182566911963
}
```
Updating my comment, I just realized that the default maximum number of steps mentioned in the paper is 1000.
Sorry, I will test it with the same configuration mentioned in the paper and share the result soon.

**Comment by hyukkyukang** [comment]
Hello,

I would like to share my evaluation result with maximum step set to 1000.
```
{
    "em": 0.6465901417960838,
    "f1": 0.7760066926190201,
    "prec": 0.8015007441843505,
    "recall": 0.7901450641766213
}
```

The accuracy did increase, but it's still approximately 3% below what is reported in the Table 2 of the paper.
I've used training data from [Google drive](https://drive.google.com/drive/folders/1n38_RIhz5jyZQUFLkp-QUVERhAI7E6nn?usp=sharing) and the training [command](https://github.com/zycdev/AISO/blob/main/train_union.py#L55) from train_union.py.

@zycdev,
Did you get the results of Table 2 with the same training data and the training command?
If so, could you please share the model checkpoint, as I am unable to reproduce the result?

Thank you for the help!

**Comment by zycdev** [comment]
Hello,

I'm sure the training data is from [Google Drive](https://drive.google.com/drive/folders/1n38_RIhz5jyZQUFLkp-QUVERhAI7E6nn?usp=sharing), but I can't find which is the best training script I've ever used. My project was scattered on several servers at that time. After several server poisoning and system reinstallations, the files I backed up were very messy. So sorry, as I haven't reorganized this project for a long time, I can't provide you with training scripts and model checkpoints quickly. But I'd be more than happy to help with your reproduction since you've understood my code and are close to making it happen.

The following are points that I think can continue to be improved:
1. As I mentioned in my paper, after some epochs of joint training, the encoders of the belief and policy modules will not be shared and they will be trained based on this common initial point for specialization. I found that the GitHub version of the code doesn't seem to support this, so this may require a bit of modification on your part.
2. The implementation of the retrievers can also have a big impact on performance. Since the index files of ElasticSearch and MDR are too large, I can't share my retrievers. But if you can provide more detailed evidence retrieval performance, I might be able to help you analyze the problem.
3. The [training data](https://drive.google.com/drive/folders/1n38_RIhz5jyZQUFLkp-QUVERhAI7E6nn?usp=sharing) is generated based on my retrievers, so the learned policy may not be well suited to the retrievers you implement. Therefore, you can generate training data more suitable for your retrievers according to [gen_step_data.ipynb](https://github.com/zycdev/AISO/blob/main/gen_step_data.ipynb) and [refine_hard_neg.ipynb](https://github.com/zycdev/AISO/blob/main/refine_hard_neg.ipynb).
4. For checkpoint selection, the proxy single-step metrics are not particularly reliable. I remember that I actually measured many checkpoints in game.ipynb.
5. By the way, the corpus file in Google Drive seems to be corrupted,[here](https://mailsucaseducn-my.sharepoint.com/:f:/g/personal/zhuyunchang17_mails_ucas_edu_cn/EhHrDX99GltLp-Ft9rUs9GsB0ROJXZ9bnDxLoLKE3MUFmg) is the reuploaded file.

Thank you for your interest!

**Comment by hyukkyukang** [comment]
Thanks for the help! 

I am currently going through each of the recommendations you suggested. I will share the results as soon as they are ready.

Besides, I have one more question regarding the corpus file you provided.
There are approximately 5.2 million documents in the [wikipedia dump](https://hotpotqa.github.io/wiki-readme.html) from the HotpotQA datasets (i.e. October 1st, 2017 dump).
However, the [corpus file](https://mailsucaseducn-my.sharepoint.com/:f:/g/personal/zhuyunchang17_mails_ucas_edu_cn/EhHrDX99GltLp-Ft9rUs9GsBRetm4PAqCqCt8nZfmwQOFA?e=HdTejM) you shared is consisted of 533,991 documents.
Can you please explain why the number is different? And how are the documents filtered from the original Wikipedia dump?
I could not find the script that generates the corpus file in the repository.

**Comment by zycdev** [comment]
That's a good question, and it made me realize that the uploaded index_sparse.ipynb was legacy. [index_sparse.5.ipynb](https://github.com/zycdev/AISO/blob/main/index_sparse.5.ipynb) is the final version of my corpus processing notebook, in which you should find the answer. 

By the way, refine_hard_neg.ipynb is also crucial for training data distribution, I have updated my last [comment](https://github.com/zycdev/AISO/issues/4#issuecomment-1541298203).

**Comment by hyukkyukang** [comment]
Before testing with other suggestions, I've tested the model with the corpus file you reshared ([here](https://mailsucaseducn-my.sharepoint.com/:f:/g/personal/zhuyunchang17_mails_ucas_edu_cn/EhHrDX99GltLp-Ft9rUs9GsBRetm4PAqCqCt8nZfmwQOFA?e=HdTejM)). I was able to see a noticeable improvement in accuracy by approximately 1.8%. The specific metrics are as follows:
```json
{
    "em": 0.6545577312626604,
    "f1": 0.7857827426294238,
    "prec": 0.811630128196888,
    "recall": 0.7992938836968118
}
```

However, I have one question regarding [index_spare.5.ipynb](https://github.com/zycdev/AISO/blob/main/index_sparse.5.ipynb).
When I run this script, it processes and saves around 5.2 million documents. This result is noticeably different from the corpus file you provided ([here](https://mailsucaseducn-my.sharepoint.com/:f:/g/personal/zhuyunchang17_mails_ucas_edu_cn/EhHrDX99GltLp-Ft9rUs9GsBRetm4PAqCqCt8nZfmwQOFA?e=HdTejM)), which contains only 0.5 million documents.

To ensure that I'm accurately replicating your results, could you clarify which of these document sets was used for Table 2 in your paper? Your clarification will greatly assist my ongoing evaluation.

**Comment by zycdev** [comment]
- `enwiki-20171001-paragraph-5` is the Wikipedia corpus where each document contains all its paragraphs.
- [`hotpot-paragraph-5`](https://mailsucaseducn-my.sharepoint.com/:u:/g/personal/zhuyunchang17_mails_ucas_edu_cn/EdcQpjPWuY1LgfHcz5IkkBMB1bNKcb1MpT1mJUbL3Y-dTA?e=UEicJ5) is the fullwiki corpus for HotpotQA where each document only contains the abstract paragraph (e.g. the first paragraph). It is a subset of `enwiki-20171001-paragraph-5`, and will be used for inference in `game.ipynb`.
- [`hotpot-paragraph-5.strict.refined`](https://mailsucaseducn-my.sharepoint.com/:u:/g/personal/zhuyunchang17_mails_ucas_edu_cn/EU5nOZzsMbtJmfsL20Szlm8Byr0CFzAkTjjAX27LbnQm_g?e=7UdI3i) is a filtered version of the fullwiki corpus. It keeps only the positive and hard negative paragraphs used in `hotpot-step-(train|dev).strict.refined.jsonl` to save memory during training.
